src/bootstrap/src/core/config/config.rs

```diff
@@ -1605,11 +1623,56 @@ impl Config {
             } = rust;
 
             is_user_configured_rust_channel = channel.is_some();
-            set(&mut config.channel, channel);
+            set(&mut config.channel, channel.clone());
 
             config.download_rustc_commit = config.download_ci_rustc_commit(download_rustc);
 
-            // FIXME: handle download-rustc incompatible options.
+            // There are two kinds of checks for CI rustc incompatible options:
+            //    1. Checking an option that may change the compiler behaviour/output.
+            //    2. Checking an option that have no effect on the compiler behaviour/output.
+            //
+            // If the option belongs to the first category, we call `err_incompatible_ci_rustc_option` macro
+            // for a hard error; otherwise, we just print a warning with `warn_incompatible_ci_rustc_option` macro.
+            if config.download_rustc_commit.is_some() {
+                err_incompatible_ci_rustc_option!(optimize_toml, "optimize");
+                err_incompatible_ci_rustc_option!(debug_toml, "debug");
+                err_incompatible_ci_rustc_option!(codegen_units, "codegen-units");
+                err_incompatible_ci_rustc_option!(codegen_units_std, "codegen-units-std");
```

Comment:
Are these really all incompatible? I'd have assumed that you still get to use all of these for a full build (i.e., basically stage 2 build). Those options related to `std` definitely feel like they still are fine -- they would certainly apply to the standard library built by the downloaded rustc, right?

Reply:
Building std from source is not currently allowed with download-rustc: https://github.com/rust-lang/rust/blob/51917ba8f215f76e9d3fa8e77cd0a781bb28dab7/src/bootstrap/src/core/build_steps/compile.rs#L155-L171

But it will eventually be possible with [redesign stage 0 std](https://github.com/rust-lang/rust/pull/119899), so I will remove the std-related options from the list.